add migration recipe to docs (#278)
python/psf-salt: docs/migration-recipe.md

Migrating to new host:
----------------------

> ### Prepare for migration:  

Reviewer: I think we should specify or make clear that this should be done locally.

Perhaps the title of this section should be "Prepare salt configuration for migration".

> ### Prepare for migration:  

Reviewer: I think we should specify or make clear that this should be done locally.

Perhaps the title of this section should be "Prepare salt configuration for migration".

1.  Ensure that salt-master, loadbalancer, and host in question are up, and that their health check for the relevant service is failing in haproxy after the host is fully up

To view haproxy status: 

-vagrant up the salt-master, loadbalancer, and host in question (`vagrant up`)

> -`vagrant ssh-config salt-master loadbalancer >> vagrant-ssh Ssh -L 4646:127:0.0.1:464 -F vagrant-ssh loadbalancer`

Reviewer: Describe what this is doing. I also think there are two commands here, let's break those up!

Something like:

```
Prepare an ssh configuration file for accessing the hosts with native ssh commands rather than vagrant ssh:
`vagrant ssh-config salt-master loadbalancer >> vagrant-ssh`

Open an ssh session with port forwarding to the haproxy status page:
`Ssh -L 4646:127:0.0.1:464 -F vagrant-ssh loadbalancer`
```

> -open loacalhost:4646/haproxy?stats in browser

Reviewer: Describe this? "View the haproxy status page in your browser: `command`"

2.  Edit pillar data for roles.sls to include both old and new hostnames (ex. hostname*)

### Migrate the host:

#### Update Salt Master with latest config including prep from above

1.  ssh into the salt-master server `ssh salt-master`

2.  Navigate to srv/psf-salt `cd srv/psf-salt`

> 3.  Run `git pull`

Reviewer: Probably needs to be `sudo git pull`, since our users don't have write access directly.

> 4.  Run `sudo salt-call state.highstate`

Reviewer: Describe why? Primarily we need the roles settings to be updated to reflect the new matching pattern! We also need any changes to support the migration available.

#### Ensure new configuration doesn't impact host being migrated

> 1.  ssh into the host in question `ssh hostname`

Reviewer: I think we can specify "old host" rather then "the host in question".

This brings to mind, that we might want to have consistency in the whole document between "old host", "new host", and "salt server".

I'll look for this in the next pass of review!

> 2.  Navigate to srv/psf-salt `cd srv/psf-salt`

Reviewer: I don't think steps 2 and 3 here are relevant to hosts that aren't the salt sever, are they?

3.  Run `git pull`

> 4.  Run `sudo salt-call state.highstate`

Reviewer: Might add a step to verify the service is still healthy on the old host!

Author: Might need a little help in understanding what this change might look like 🙃

#### Create new host:

1.  Start a new droplet  in digital ocean, and check resources being used on old host to see if we are over or under spending on resources 

2.  Create a new droplet with new version of ubuntu, appropriate resources,  and name it according to hostname + 2004

#### Provision new host for migration:

1.  Add Salt repositories for our current target version (add the apt-repo and install salt-minion package)

> -  `RUN wget --quiet -O /usr/share/keyrings/salt-archive-keyring.gpg https://repo.saltproject.io/py3/ubuntu/20.04/$(dpkg --print-architecture)/3004/salt-archive-keyring.gpg`

Reviewer: `RUN` is specific to Dockerfile, We can just start with `wget`.

> -  `RUN echo "deb [signed-by=/usr/share/keyrings/salt-archive-keyring.gpg arch=$(dpkg --print-architecture)] https://repo.saltproject.io/py3/ubuntu/20.04/$(dpkg --print-architecture)/3004 focal main" > /etc/apt/sources.list.d/salt.list`

Reviewer: `RUN` is specific to Dockerfile, We can just start with `echo`.

> 2.  Configure the salt-minion. On new host, run the command

Reviewer: I think you mean Install and configure the salt-minion package

> - `RUN apt-get update -y && apt-get install -y --no-install-recommends salt-minion`

Reviewer: (same note on `RUN`)

- On the original host, look through `etc/salt/minion.d*` to set up salt-minion configuration files to match on new host 

> 3.  Run `sudo salt-call state.highstate`

Reviewer: Unfortunately, I don't think that high state will succeed at this point.

We're missing a few steps.

Namely,

- Restart the salt-minion service on the new host to pickup the configuration and register with the salt-master
- Accept the key for the new host on the salt-master

> #### Stop services on old host:

Reviewer: I don't think we want to stop the services on the old host, until the bulk of our data has been moved by the rsync command on line 80.

1.  ssh into old host ( `ssh old-hostname` )

2.  stop cron jobs, `sudo service cron stop`

3.  stop public-facing services, like nginx, or the service the health check is looking for ex)  `sudo service nginx stop`

#### Begin data migration:

1.  `ssh -A new-hostname` into new host to enable forwarding of ssh-agent

2.  stop cron jobs, `sudo service cron stop`

3.  stop public-facing services, like nginx, or the service the health check is looking for ex)  `sudo service nginx stop`

4.  ensure that the volume is mounted in the correct location 

> 5.  run rsync once to move bulk of data and as necessary to watch for changes, and one final time before making the swap `sudo -E -s rsync -avz --rsync-path="sudo rsync" username@hostname: /pathname/ /pathname/` 

Reviewer: This step could probably be broken up in to a  couple steps:

- Bulk data transfer
- Check data transfer (see how many changes there are)
- Shutdown cron/services on old host
- Final data transfer

1.  The `/pathname/` can be determined by looking at the pillar data for backups, `pillar/prod/backup` using the source_directory path for the given host (example: the downloads host uses `/srv/`)

1.  start cron jobs, `sudo service cron start`

2.  start public-facing services involved with healthcheck, like nginx, `sudo service nginx start`

3.  check if users have any files on old host and transfer accordingly

#### Shutdown and reclaim hostname:

1.  On old host, destroy old host by running, `sudo -h shutdown now`

2.  Change the hostname in Digital Ocean 

3.  On new host, run `sudo hostname new-hostname` to rename

> 4.  Update hostname in `etc/hostname`,  `etc/salt/minion_id`, and `etc/hosts`

Reviewer: I think we need to restart the salt-minion after editing the minion_id file

> 5.  Run `sudo salt-key -L` to list out and remove old keys 

Reviewer: `salt-key -L` only lists keys, we need `salt-key -d hostname` to delete the old key.

Migrating to new host:
----------------------

### Prepare salt configuration for migration:  

> 1.  Ensure that salt-master, loadbalancer, and host in question are up, and that their health check for the relevant service is failing in haproxy after the host is fully up

Reviewer: Could add bulleted list of commands to run:

```
vagrant up salt-master
vagrant up loadbalancer
vagrant up host
```

> 1.  Ensure that salt-master, loadbalancer, and host in question are up, and that their health check for the relevant service is failing in haproxy after the host is fully up

Reviewer: Could add bulleted list of commands to run:

```
laptop:psf-salt user$ vagrant up salt-master
laptop:psf-salt user$ vagrant up loadbalancer
laptop:psf-salt user$ vagrant up host
```

we might also want to make it explicit that we are doing this locally.  so instead of "Ensure that salt-master, loadbalancer, and host in question are up" Maybe "Ensure that salt-master, loadbalancer, and host in question can be brought up with vagrant locally"

To view haproxy status: 

> -vagrant up the salt-master, loadbalancer, and host in question (`vagrant up`)

Reviewer: if you correct the markdown formatting to have a space after the leading `-` here and on the next few lines you should get a pretty list!

-Prepare an ssh configuration file to access the host with native ssh commands: `vagrant ssh-config salt-master loadbalancer >> vagrant-ssh` 

> -Open an ssh session with port forwarding to the haproxy status page: `ssh -L 4646:127:0.0.1:464 -F vagrant-ssh loadbalancer`

Reviewer: ```suggestion
-Open an ssh session with port forwarding to the haproxy status page: `ssh -L 4646:127.0.0.1:4646 -F vagrant-ssh loadbalancer`
```

> -view the haproxy status page in your browser `loacalhost:4646/haproxy?stats` 

Reviewer: ```suggestion
-view the haproxy status page in your browser `http://localhost:4646/haproxy?stats` 
```

> 2.  Edit pillar data for roles.sls to include both old and new hostnames (ex. hostname*)

Reviewer: We could show a sample diff here of an actually change!

```
diff --git a/pillar/prod/roles.sls b/pillar/prod/roles.sls
index 68387c9..7a8ace1 100644
--- a/pillar/prod/roles.sls
+++ b/pillar/prod/roles.sls
@@ -35,7 +35,7 @@ roles:
     purpose:  "Builds and serves CPython's documentation"
     contact:  "mdk"
   downloads:
-    pattern:  "downloads.nyc1.psf.io"
+    pattern:  "downloads*.nyc1.psf.io"
     purpose:  "Serves python.org downloads"
     contact:  "CPython Release Managers"
   hg:
```

### Migrate the host:

#### Update Salt Master with latest config including prep from above

> 1.  ssh into the salt-master server `ssh salt-master`

Reviewer: ```suggestion
1.  ssh into the salt-master server `ssh salt.nyc1.psf.io`
```

> 2.  Navigate to srv/psf-salt `user@salt:~$ cd srv/psf-salt`

Reviewer: ```suggestion
2.  Navigate to srv/psf-salt `user@salt:~$ cd /srv/psf-salt`
```

3.  Run `user@salt:/srv/psf-salt$ sudo git pull`

4.  Run highstate to update the roles settings to reflect the new matchng pattern, as well as additional changes to support migration: `user@salt:/srv/psf-salt$ sudo salt-call state.highstate`

#### Ensure new configuration doesn't impact host being migrated

> 1.  ssh into the old host `ssh hostname`

Reviewer: Let's make sure we're keeping consistent with the example host names, so I think this line might want to be:

"1.  ssh into the old-host `ssh old-host`"

2.  Run `user@old-host:~$ sudo salt-call state.highstate`

#### Create new host:

1.  Start a new droplet  in digital ocean, and check resources being used on old host to see if we are over or under spending on resources 

2.  Create a new droplet with new version of ubuntu, appropriate resources,  and name it according to hostname + 2004

#### Provision new host for migration:

> 1. ssh into new host `ssh new-host`

Reviewer: We'll need to ssh in for the first time via IP address, since the hostname won't be created by salt yet.

Suggest: "1. ssh into the new-host via the IP address provided by DigitalOcean `ssh root@NNN.NNN.NNN.NNN``

2.  Add Salt repositories for our current target version (add the apt-repo and install salt-minion package)

-   `user@new-host:~$ wget --quiet -O /usr/share/keyrings/salt-archive-keyring.gpg https://repo.saltproject.io/py3/ubuntu/20.04/$(dpkg --print-architecture)/3004/salt-archive-keyring.gpg`

-  `user@new-host:~$ echo "deb [signed-by=/usr/share/keyrings/salt-archive-keyring.gpg arch=$(dpkg --print-architecture)] https://repo.saltproject.io/py3/ubuntu/20.04/$(dpkg --print-architecture)/3004 focal main" > /etc/apt/sources.list.d/salt.list`

3.  Install and configure the salt-minion. On new host, run the command

- `user@new-host:~$ apt-get update -y && apt-get install -y --no-install-recommends salt-minion`

- On the original host, look through `etc/salt/minion.d*` to set up salt-minion configuration files to match on new host 

4. Restart the salt-minion service on the new host to pickup the configuration and register with salt-master: `user@new-host:~$ sudo salt-call service.restart`

5.  On salt-master, accept the key for the new host: `user@new-host:~$ sudo salt-key -a hostname`

> 6.  Run `user@new-host:~$ sudo salt-call state.highstate`

Reviewer: An important check at this point is to ensure that the new host _is not_ serving traffic since no data has been migrated.

I suggest a 7th step of:

"7. Ensure that the new host is not passing health checks in the loadbalancer: `ssh -L 4646:127.0.0.1:4646 lb-a.nyc1.psf.io` then open http://localhost:4646/haproxy?stats in your browser.

One other thing to "finalize" a new host, is to make sure that the salt-master has run highstate after the the new-host has finished a high state for the first time.

I suggest a 8th step, of :

"8. Run highstate on the salt-master to create public dns record for the new-host `user@salt:/srv/psf-salt$ sudo salt-call state.highstate`"

3.  Install and configure the salt-minion. On new host, run the command

- `user@new-host:~$ apt-get update -y && apt-get install -y --no-install-recommends salt-minion`

- On the original host, look through `etc/salt/minion.d*` to set up salt-minion configuration files to match on new host 

4. Restart the salt-minion service on the new host to pickup the configuration and register with salt-master: `user@new-host:~$ sudo salt-call service.restart`

5.  On salt-master, accept the key for the new host: `user@new-host:~$ sudo salt-key -a hostname`

6. Ensure that the new host is not passing health checks in the loadbalancer: `ssh -L 4646:127.0.0.1:4646 lb-a.nyc1.psf.io` then open http://localhost:4646/haproxy?stats in your browser.
 
7.  Run hightstate on the salt-master to create a public dns record for the new-host `user@salt:/srv/psf-salt$ sudo salt-call state.highstate`

#### Begin data migration:

1.  `ssh -A new-host` into new host to enable forwarding of ssh-agent

2.  stop cron jobs, `user@new-host:~$ sudo service cron stop`

3.  stop public-facing services, like nginx, or the service the health check is looking for ex)  `user@new-host:~$ sudo service nginx stop`

> 4.  ensure that the volume is mounted in the correct location 

Reviewer: Not all hosts will have extra volumes attached, we may say "ensure that any additional volumes are mounted and in the correct location"?

Reviewer: Could add details like

- Check what disks are currently mounted and where: `df`
- Determine where any additional disks should be mounted (based on salt configuration of services, for example `docs` and `downloads` roles needs a big `/srv` for their data storage.
- Ensure mounting of any external disks in the right location using `mount` command with appropriate arguments
- Ensure that the volumes will be remounted on startup by configuring them in `/etc/fstab`

> 5.  run rsync once to move bulk of data and as necessary to watch for changes `user@new-host:~$ sudo -E -s rsync -avz --rsync-path="sudo rsync" username@hostname: /pathname/ /pathname/` 

Reviewer: ```suggestion
5.  run rsync once to move bulk of data and as necessary to watch for changes `user@new-host:~$ sudo -E -s rsync -av --rsync-path="sudo rsync" username@old-host:/pathname/ /pathname/` 
```

Reviewer: Two changes, a typo in the host:/path/ part of the command (spaces are an error) and removing unnecessary compression step for the sync since these hosts are very likely to be on solid/fast internet.

> -The `/pathname/` can be determined by looking at the pillar data for backups, `pillar/prod/backup` using the source_directory path for the given host (example: the downloads host uses `/srv/`)

Reviewer: ```suggestion
- The `/pathname/` can be determined by looking at the pillar data for backups, `pillar/prod/backup` using the source_directory path for the given host (example: the downloads host uses `/srv/`)
```

Reviewer: Suggest adding this space so that the list will be rendered pretty! I won't note this on any remaining spots I find, might be good to scan for similar issues that keep lists from rendering.

#### Stop services on old host:

1.  ssh into old host ( `ssh old-host` )

2.  stop cron jobs, `user@old-host:~$ sudo service cron stop`

3.  stop public-facing services, like nginx, or the service the health check is looking for ex)  `user@old-host:~$ sudo service nginx stop`

#### Finish data migration and restart cron/public-facing services:

> 1. run rsync once more to finalize data migration `user@new-host:~$ sudo -E -s rsync -avz --rsync-path="sudo rsync" username@hostname: /pathname/ /pathname/` 

Reviewer: ```suggestion
1. run rsync once more to finalize data migration `user@new-host:~$ sudo -E -s rsync -avz --rsync-path="sudo rsync" username@old-host:/pathname/ /pathname/` 
```

2.  start cron jobs, `user@new-host:~$ sudo service cron start`

> 3.  start public-facing services involved with healthcheck, like nginx, `user@new-host:~$ sudo service nginx start`

Reviewer: I think we want a step after this to ensure that the new host is live an serving traffic. By either checking the loadbalancer page or any associated public services.

> 4.  check if users have any files on old host and transfer accordingly

Reviewer: We can and should write a command for this... I think it'd look something like:

```
user@new-host:~$ for user in /home/psf-users/*; do sudo -E -s rsync --delete -av --progress --rsync-path="sudo rsync" user@old-host:$user/ $user/migrated-from-ubuntu-1804-lts-host/; done
```

> #### Shutdown and reclaim hostname:

Reviewer: All through this section, let's make sure we're being consistent with `new-host` and `old-host`.

> 1.  On old host, destroy old host by running, `user@old-host:~$ sudo -h shutdown now`

Reviewer: ```suggestion
1.  On old host, stop the old host by running, `user@old-host:~$ sudo -h shutdown now`
```

> 

Reviewer: I think we need a step here to destroy the old host in DigitalOcean. Shutting down just stops it. It isn't _strictly_ required. Alternatively we could just rename the stopped server prefixing with something like `old-` if we want to be safe.

> 2.  Change the hostname in Digital Ocean 

Reviewer: of which host? presuming its the new one, we can specify "to remove the suffix or similar that was used to differentiate it from the old host".

3.  On new host, run `user@new-host:~$ sudo hostname new-host` to rename

> 4.  Update hostname in `etc/hostname`,  `etc/salt/minion_id`, and `etc/hosts`

Reviewer: I think all these file paths need a leading `/`
